### PR TITLE
Jailed node should be unstaked if enough validators

### PR DESCRIPTION
**Pull request by raduchis**
Fix for jailed validators when the queue size is 0.

Before the fix: a jailed validator would remain staked and if a node was added to the queue, it wouldn't take the jailed node's place
After the fix: a jailed validator is removed from the staked nodes and a node that is added to the queue will get staked (status new) immediately.

### CODE REVIEW

File: vm/systemSmartContracts/staking.go [hunk 1]
```diff
@@ -116,31 +118,33 @@ func NewStakingSmartContract(
...
+		validatorToDelegationEnableEpoch:        args.EpochConfig.EnableEpochs.ValidatorToDelegationEnableEpoch,
+		correctFirstQueuedEpoch:                 args.EpochConfig.EnableEpochs.CorrectFirstQueuedEpoch,
+		correctJailedNotUnstakedEmptyQueueEpoch: args.EpochConfig.EnableEpochs.CorrectJailedNotUnstakedEmptyQueueEpoch,
 	}
 	log.Debug("staking: enable epoch for stake", "epoch", reg.enableStakingEpoch)
 	log.Debug("staking: enable epoch for staking v2", "epoch", reg.stakingV2Epoch)
 	log.Debug("staking: enable epoch for correct last unjailed", "epoch", reg.correctLastUnjailedEpoch)
 	log.Debug("staking: enable epoch for validator to delegation", "epoch", reg.validatorToDelegationEnableEpoch)
 	log.Debug("staking: enable epoch for correct first queued", "epoch", reg.correctFirstQueuedEpoch)
+	log.Debug("staking: enable epoch for correct jailed not unstaked with empty queue", "epoch", reg.flagCorrectJailedNotUnstakedEmptyQueue)
```
Comment (sasurobert) [inline vm/systemSmartContracts/staking.go:147]: you need to add log in node/nodeRunner.go as well

File: vm/systemSmartContracts/staking.go [hunk 2]
```diff
@@ -1196,6 +1198,26 @@ func (s *stakingSC) switchJailedWithWaiting(args *vmcommon.ContractCallInput) vm
 	return vmcommon.Ok
 }
 
+func (s *stakingSC) tryRemoveJailedNodeFromStaked(registrationData *StakedDataV2_0) {
+	if !s.flagCorrectJailedNotUnstakedEmptyQueue.IsSet() {
+		s.removeAndSetUnstaked(registrationData)
+	} else {
```
Comment (sasurobert) [inline vm/systemSmartContracts/staking.go:1204]: add a return here. and there is no need for else

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -1196,6 +1198,26 @@ func (s *stakingSC) switchJailedWithWaiting(args *vmcommon.ContractCallInput) vm
...
+		if s.canUnStake() {
+			s.removeAndSetUnstaked(registrationData)
```
Comment (sasurobert) [inline vm/systemSmartContracts/staking.go:1206]: add return here and delete else

File: vm/systemSmartContracts/staking.go (hunk 1 above)
Comment (raduchis) [reply vm/systemSmartContracts/staking.go:147]: fixed

File: vm/systemSmartContracts/staking.go (hunk 2 above)
Comment (raduchis) [reply vm/systemSmartContracts/staking.go:1204]: fixed

File: vm/systemSmartContracts/staking.go (hunk 3 above)
Comment (raduchis) [reply vm/systemSmartContracts/staking.go:1206]: fixed

File: vm/systemSmartContracts/staking.go (hunk 1 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/staking.go:147]: req.correctJailedNotUnstakedEmptyQueueEpoch instead reg.flagCorrectJailedNotUnstakedEmptyQueue

File: vm/systemSmartContracts/staking.go [hunk 4]
```diff
@@ -2087,6 +2111,9 @@ func (s *stakingSC) EpochConfirmed(epoch uint32, _ uint64) {
 
 	s.flagCorrectFirstQueued.Toggle(epoch >= s.correctFirstQueuedEpoch)
 	log.Debug("stakingSC: correct first queued", "enabled", s.flagCorrectFirstQueued.IsSet())
+
+	s.flagCorrectJailedNotUnstakedEmptyQueue.Toggle(epoch >= s.correctFirstQueuedEpoch)
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:2115]: s.correctJailedNotUnstakedEmptyQueueEpoch instead s.correctFirstQueuedEpoch

Comment (raduchis) [reply vm/systemSmartContracts/staking.go]: fixed

**Review by SebastianMarian — DISMISSED** [review]

**Review by sasurobert — DISMISSED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Pull request merged**